csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs

```diff
@@ -102,55 +103,145 @@ public static ExitCode Run(string[] args)
+                return ExitCode.Errors;
+            }
+        }
 
-                var compilerArguments = CSharpCommandLineParser.Default.Parse(
-                    compilerVersion.ArgsWithResponse,
-                    workingDirectory,
-                    compilerVersion.FrameworkPath,
-                    compilerVersion.AdditionalReferenceDirectories
-                    );
+        private static ExitCode RunBinaryLogAnalysis(Stopwatch stopwatch, Options options, string binlogPath, ILogger logger, CanonicalPathCache canonicalPathCache, PathTransformer pathTransformer)
+        {
+            logger.LogInfo($"Reading compiler calls from binary log {binlogPath}");
+            try
+            {
+                using var fileStream = new FileStream(binlogPath, FileMode.Open, FileAccess.Read, FileShare.Read);
+                using var reader = BinaryLogReader.Create(fileStream);
 
-                if (compilerArguments is null)
+                // Filter out compiler calls that aren't interesting for examination
+                static bool filter(CompilerCall compilerCall)
                 {
-                    var sb = new StringBuilder();
-                    sb.Append("  Failed to parse command line: ").AppendList(" ", compilerArgs);
-                    logger.Log(Severity.Error, sb.ToString());
-                    ++analyser.CompilationErrors;
-                    return ExitCode.Failed;
+                    return compilerCall.IsCSharp &&
+                        compilerCall.Kind == CompilerCallKind.Regular;
                 }
 
-                if (!analyser.BeginInitialize(compilerVersion.ArgsWithResponse))
+                var allCompilationData = reader.ReadAllCompilationData(filter);
+                var allFailed = true;
+
+                logger.LogInfo($"  Found {allCompilationData.Count} compilations in binary log");
+
+                foreach (var compilationData in allCompilationData)
                 {
-                    logger.Log(Severity.Info, "Skipping extraction since files have already been extracted");
-                    return ExitCode.Ok;
+                    if (compilationData.GetCompilationAfterGenerators() is not CSharpCompilation compilation)
+                    {
+                        logger.LogError("  Compilation data is not C#");
+                        continue;
+                    }
+
+                    var compilerCall = compilationData.CompilerCall;
+                    var diagnosticName = compilerCall.GetDiagnosticName();
+                    logger.LogInfo($"  Processing compilation {diagnosticName}");
+                    var compilerArgs = compilerCall.GetArguments();
+                    var args = reader.ReadCommandLineArguments(compilerCall);
+
+                    // Generated syntax trees are always added to the end of the list of syntax trees.
+                    var generatedSyntaxTrees = compilation.SyntaxTrees.Skip(compilationData.Compilation.SyntaxTrees.Count());
```

Comment:
Relying on the order seems somewhat fragile. Should we instead use a set subtraction?

Reply:
Maybe create a hashset of the `ompilationData.Compilation.SyntaxTrees` - it seems like that the same object references are used in the collection that also contains the syntax trees after generators are run.

Reply:
I haven't decided yet. @jaredpar what do you think? Can we rely on complog to always add generated ASTs to the end of the list?

Reply:
> Can we rely on complog to always add generated ASTs to the end of the list?

That is how the compiler works today and compiler logs emulate that behavior. 

While putting the generated trees last is an implementation detail of the compiler, it's one that is _very unlikely_ to ever change. File ordering can impact C# semantics in a few cases (field initialization in `partial` types, determinism, etc ...) hence any ordering change would be a potential breaking change. Very unlikely we'd ever do that.

At the same time I understand it feels weird to depend on that is source like this. You could call `compilationData.GetGeneratedSyntaxTrees` instead. Under the hood it does the same thing but in the case the compiler did ever change this method would also change.